src/CommandAction.ts

```diff
@@ -0,0 +1,108 @@
+import { commands, window, Selection, TextEditor } from "vscode";
+import {
+  Action,
+  ActionPreferences,
+  ActionReturnValue,
+  Graph,
+  TypedSelection,
+  SelectionWithEditor,
+} from "./Types";
+import displayPendingEditDecorations from "./editDisplayUtils";
+import { runOnTargetsForEachEditor } from "./targetUtils";
+import { focusEditor } from "./setSelectionsAndFocusEditor";
+import { flatten } from "lodash";
+
+export default class CommandAction implements Action {
+  targetPreferences: ActionPreferences[] = [{ insideOutsideType: "inside" }];
+
+  constructor(private graph: Graph, private command: string) {
+    this.run = this.run.bind(this);
+  }
+
+  async run([targets]: [
+    TypedSelection[],
+    TypedSelection[]
+  ]): Promise<ActionReturnValue> {
+    await displayPendingEditDecorations(
+      targets,
+      this.graph.editStyles.referenced,
+      this.graph.editStyles.referencedLine
+    );
+
+    const originalEditor = window.activeTextEditor;
+
+    const thatMark = flatten(
+      await runOnTargetsForEachEditor<SelectionWithEditor[]>(
+        targets,
+        async (editor, selections) => {
+          // For command to the work we have to have the correct editor focused
+          if (editor !== window.activeTextEditor) {
+            await focusEditor(editor);
+          }
+
+          const originalSelections = editor.selections;
+
+          const newSelections = selections.map(
+            (selection) => selection.selection.selection
+          );
+
+          editor.selections = newSelections;
+
+          await commands.executeCommand(this.command);
+
+          const lineDeltas = calculateLineChanges(editor, newSelections);
+
+          const updateSelection = (selection: Selection) =>
```

Comment:
It just calculates a map of all the character index changes per line as a zero index based delta. The reason why we can't just use the current selection is because after an indentation where the first token of the line is selected the indentation becomes part of the selection. In other words the first token selection begins with the indentation.

Reply:
Comment added.